### PR TITLE
Add repository file resource

### PR DESCRIPTION
**Pull request by shoekstra**
Hello,

This PR adds support to manage files in a git repo using the [contents API](https://developer.github.com/v3/repos/contents).  I think I've got most cases covered in tests and have updated the docs and README appropriately.

Note that certain CI tests will be skipped until the new `GITHUB_TEST_USER_NAME` and `GITHUB_TEST_USER_EMAIL` variables are exported.

Stephen

### CODE REVIEW

**Comment by shoekstra** [comment]
For some additional context regarding the "Fix go deps" commit: when using go 1.13 I received the following:

```
go: golang.org/x/oauth2@v0.0.0-20190604054615-0f29369cfe45: invalid pseudo-version: does not match version-control timestamp (2019-06-04T05:34:49Z)
```

I resolved this by following the steps I found at https://tip.golang.org/doc/go1.13#version-validation

**Comment by kpfleming** [comment]
That change to `go.mod` is already present in #310 (on its own) and some other PRs (bundled with other changes). I would remove that change from this PR so it can be handled on its own.

File: github/resource_github_repository_file.go [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+			"sha": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The blob SHA of the file",
+			},
+		},
+	}
+}
+
+func resourceGithubRepositoryFileOptions(d *schema.ResourceData) (*github.RepositoryContentFileOptions, error) {
```
Comment (kpfleming) [inline github/resource_github_repository_file.go]: Nicely done here, good feedback to the user if they've configured the resource improperly.

File: website/docs/r/repository_file.html.markdown [hunk 2]
```diff
@@ -0,0 +1,64 @@
+---
+layout: "github"
+page_title: "GitHub: github_repository_file"
+description: |-
+  Creates and manages files within a GitHub repository
+---
+
+# github_repository_file
+
+This resource allows you to create and manage files within your
```
Comment (kpfleming) [inline website/docs/r/repository_file.html.markdown]: s/your/a/ to match the description above

**Comment by shoekstra** [comment]
> That change to `go.mod` is already present in #310 (on its own) and some other PRs (bundled with other changes). I would remove that change from this PR so it can be handled on its own.

No problem, I've dropped that commit 👍

File: github/resource_github_repository_file.go [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+
+func resourceGithubRepositoryFile() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceGithubRepositoryFileCreate,
+		Read:   resourceGithubRepositoryFileRead,
+		Update: resourceGithubRepositoryFileUpdate,
+		Delete: resourceGithubRepositoryFileDelete,
+		Importer: &schema.ResourceImporter{
+			State: func(d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+				parts := strings.Split(d.Id(), ":")
```
[the remark this answers is not in the document]
Comment (shoekstra) [reply github/resource_github_repository_file.go]: I considered `parseTwoPartID` until I saw that the error returned is quite specific and doesn't apply to what I'm parsing. As `parseTwoPartID` is used in a couple of other funcs I decided if I was going to refactor the returned error it probably needs it's own PR/discussion.

[the remark this answers is not in the document]
Comment (kpfleming) [reply github/resource_github_repository_file.go]: I've got a PR at #313 to address that problem too, so if you switch to it now when that is merged your code will have to be updated to provide the proper labels for the left/right halves of the ID.

[the remark this answers is not in the document]
Comment (shoekstra) [reply github/resource_github_repository_file.go]: Looks good - what's the delay with merging #313? Feels a bit odd to amend the commit to include broken code, maybe better to wait until #313 is merged and then rebase this PR?

[the remark this answers is not in the document]
Comment (shoekstra) [reply github/resource_github_repository_file.go]: Looks good - what's the delay with merging #313? Feels a bit odd to amend the commit to include broken code, maybe better to wait until #313 is merged and then rebase this PR?

**Comment by shoekstra** [comment]
Hey @jcudit,

> Pull Requests are a missing piece that I'd like to see added onto this either in a subsequent release or via a refactoring of this PR. Can you take a look at the linked issue and let me know if what was planned there aligns with your use cases that this patch solves? I am interested in your take on what next steps could be to involve the [Pull Request](https://github.com/google/go-github/blob/v28.0.1/github/pulls.go) functionality of the `go-github` client library.

Skipping the pull request bit was a conscious decision I made; I wanted a way to manage a few files (GitHub Actions in my case) in a bunch of repos from a central Terraform repo. The review would already happen when updating the Terraform plan and I decided creating additional PRs created additional overhead for someone to review the changes a second time and, more importantly, this didn’t fit with Terraform’s job of ensuring “desired state” in that a file would be described in a plan with desired content but this could not be ensured unless committing and pushing to the specified branch.

Reading the issue you linked, the initial proposal is the better way to go should you want to add support for pull requests. With Terraform you want to manage the file and update it using a pull request, you don’t want to manage a pull request as a merged pull request would be void if the content is modified (I hope that makes sense). 

I'm not convinced managing pull requests is something to do but am open to more discussion on how we could make it work if it's functionality you're after.

> Aside from that concern, here are a few questions I have around the implementation as-is:
> 
> * I see the `client.Repositories.CreateFile` function was used in `resourceGithubRepositoryFileUpdate` instead of the available `client.Repositories.UpdateFile` function.  What is the tradeoff around using one over the other?

Looking now I see they’re the same, there is no difference between them. At the time I referred to the [API docs](https://developer.github.com/v3/repos/contents/#create-or-update-a-file) and saw creating and updating files was done using the same method, so I didn’t look for an Update function in `go-github`.

> * I see that a branch needs to exist before it can receive commits.  What are your thoughts around adding support for creating a branch if it does not exist (and throwing an error if a branch already exists)?  Are you currently automating the creation of branches through some other means?

My use case is to create files in branches that already exist thus creating a branch is out of scope of this resource. If I wanted a branch to exist I would want to manage it via a `github_branch` resource and it would be a depended on by the `github_repository_file` resource. I would only create a branch via this resource if we implemented the pull request functionality.

> * Are you open to generalizing the resource from `_file` to `_content`? This would allow for declaring multiple files within a resource block and iterating through the lifecycle management of each one.  

Sure, I'm  happy to add a `files` field if it’ll be useful for others. I originally decided to keep to a single file per resource to ensure a file is only managed by one resource in a plan. I'm not sure how I would do this if I add multiple files in a single resource.

> Have there been any pain points with using multiple `repository_file` resources with your use cases so far?

I haven’t had any issues with multiple resources yet; at the moment I’ve declared 3 files in a module and reference that module 6 times in a plan, so effectively managing 18 files across 6 repos in a single Terraform run.

**Comment by jcudit** [comment]
> The review would already happen when updating the Terraform plan

👍 this did not occur to me and removes my desire to have Terraform initiate PRs.

---

Thank you for addressing all other concerns raised as well.  I think the workflow you are presenting here is valuable to the community and will work to get this merged and released as-is.  My next steps are to:

- [ ] Update the CI environment to ensure the new acceptance tests pass
- [ ] Work through broken tests with https://github.com/terraform-providers/terraform-provider-github/pull/313
- [ ] Merge a rebased version of this branch

**Comment by jcudit** [comment]
@shoekstra I've merged #313.  Can you rebase and consume the new `parseTwoPartID` function?

I will look into adding the new branch and environment variables to CI in the meantime.

**Comment by shoekstra** [comment]
I've rebased my branch. `parseTwoPartID` doesn't fit what I'm doing so no changes there.

I was thinking about your requests re UX. It's out of scope of me at the moment, but there should be enough of a base for someone to pick up adding the `pull_request` or `files` fields you mentioned.

Feel free to merge if you agree!

**Pull request merged**